### PR TITLE
Add `ns-db prune` stack subcommand

### PR DESCRIPTION
**Pull request by rvolosatovs**
#### Summary

References #2908 

#### Changes

- Add `ns prune` stack subcommand

#### Testing

Start 3.9.1, generate a bunch of uplinks, switch to latest 3.9, generate a bunch of uplinks and run the tool

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [x] Documentation: Relevant documentation is added or updated.
- [x] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

File: cmd/ttn-lw-stack/commands/redis.go [hunk 1]
```diff
@@ -0,0 +1,115 @@
+// Copyright © 2019 The Things Network Foundation, The Things Industries B.V.
```
Comment (rvolosatovs) [inline cmd/ttn-lw-stack/commands/redis.go]: will update to 2020

**Comment by KrishnaIyer** [comment]
Can you add some documentation here;
- Necessary config. I don't think we need to dump the entire `ns`  config object for this tool right?
- How often/under what conditions should this tool be run?

**Comment by rvolosatovs** [comment]
>     * Necessary config. I don't think we need to dump the entire `ns`  config object for this tool right?

It requires configuration in `redis` namespace sufficient to create a client, that's it.
Logger configuration is required if you want to see output.

>     * How often/under what conditions should this tool be run?

When updating from 3.9.1 or below to a version above that. It should not be necessary in other cases, but that may change and I intend to add more GC subcommands here. It is designed to provide minimum load on the DB, it runs slow on purpose and it can be run anytime.

**Comment by johanstokking** [comment]
This command is very specific for one use case whose is very conditional. Since we have the compatibility commitment on CLI, this is going to stay in V3 until the end of time.

We may not have application uplink queues anymore, let alone Redis.

Therefore, can we have a more general `ns-db prune` command, like we have `is-db` commands?

**Comment by rvolosatovs** [comment]
> This command is very specific for one use case whose is very conditional. Since we have the compatibility commitment on CLI, this is going to stay in V3 until the end of time.
> 
> We may not have application uplink queues anymore, let alone Redis.
> 
> Therefore, can we have a more general `ns-db prune` command, like we have `is-db` commands?

We can extract this into a separate tool, but this is specific to Redis and application uplink queues. The intention is to add more subcommands to `ns` namespace.
`ns-db prune` apart from being ugly, requires then the CLI determine what storage backend is used automatically (error prone and unnecessary) and could potentially delete a lot of things we would not want to have deleted. The idea is to have, for example, utilities to handle stuck entries in streams (i.e. delivered to a consumer, but never acked due to failure), possible lock-contention etc. In any case, the whole point here is to be specific about Redis.

**Comment by rvolosatovs** [comment]
So I'm fine with extracting this into standalone `ns-redis-tool` or whatever, but definitely not `*-lw-stack ns-db prune`

**Comment by rvolosatovs** [comment]
Can't we just make this subcommand hidden, such that customers can't depend on it?

**Comment by johanstokking** [comment]
> `ns-db prune` apart from being ugly

What is ugly about it? We have `is-db` already, and `prune` is quite common. I find it more clear from a user perspective than `gc`

> requires then the CLI determine what storage backend is used automatically

`ttn-lw-stack` already does that based on config. It doesn't matter that NS currently only uses Redis from the user's perspective, like `is-db` is also generic and is not `ttn-lw-stack is postgresql insert-user` or something.

> and could potentially delete a lot of things we would not want to have deleted

That's going to be the magic of `prune`. And we can always make confirmations or flags should that be necessary. See https://docs.docker.com/engine/reference/commandline/system_prune/

> utilities to handle stuck entries in streams [...], possible lock-contention etc

If we need CLI to recover for these specific things, we're doing something wrong. We are not going to increase ops burden to manually fix these things and document everything.

**Comment by rvolosatovs** [comment]
> What is ugly about it? We have `is-db` already, and `prune` is quite common. I find it more clear from a user perspective than `gc`

`is-db` or `ns-db` is. It should be `is db <command>` or just `is create-admin-user`, for example. The `-db` part is completely irrelevant and non-intuitive

File: cmd/ttn-lw-stack/commands/redis.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+		}
+	}
+	return nil
+}
+
+var (
+	nsCommand = &cobra.Command{
+		Use:     "ns",
+		Aliases: []string{"ns-db"},
+		Short:   "Manage Network Server",
```
Comment (htdvisser) [inline cmd/ttn-lw-stack/commands/redis.go]: You may not agree with our historical choices for naming commands, but I think we can all agree that we prefer our commands to be consistent. So please instead write:

```
Use:   "ns-db",
Short: "Manage the Network Server database",
```

**Review by htdvisser — CHANGES_REQUESTED** [review]

File: cmd/ttn-lw-stack/commands/redis.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+			}
+
+			logger.Info("Connecting to Redis database...")
+			cl := NewNetworkServerApplicationUplinkQueueRedis(*config)
+			var deleted uint64
+			defer func() { logger.Debugf("%d processed stream entries deleted", deleted) }()
+			return rangeRedisKeys(cl, nsredis.ApplicationUplinkQueueUIDGenericUplinkKey(cl, "*"), func(k string) bool {
+				gs, err := cl.XInfoGroups(k).Result()
+				if err != nil {
+					logger.WithError(err).Errorf("Failed to query groups of stream '%s'", k)
```
Comment (johanstokking) [inline cmd/ttn-lw-stack/commands/redis.go]: Isn't `gs` undefined at this point?

File: cmd/ttn-lw-stack/commands/redis.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+				}
+				for _, g := range gs {
+					if g.Name != "ns" {
+						logger.Errorf("Unexpected consumer group with name '%s' found for stream '%s'", g.Name, k)
```
Comment (johanstokking) [inline cmd/ttn-lw-stack/commands/redis.go]:
```suggestion
						logger.Errorf("Unexpected consumer group with name %q found for stream %q", g.Name, k)
```

File: cmd/ttn-lw-stack/commands/redis.go [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+						continue
+					}
+					last := "-"
+					for {
+						msgs, err := cl.XRangeN(k, last, g.LastDeliveredID, 1).Result()
+						if err != nil {
+							logger.WithError(err).Errorf("Failed to XRANGE over stream '%s'", k)
```
Comment (johanstokking) [inline cmd/ttn-lw-stack/commands/redis.go]:
```suggestion
							logger.WithError(err).Errorf("Failed to XRANGE over stream %q", k)
```

File: cmd/ttn-lw-stack/commands/redis.go [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+							return true
+						}
+						var ids []string
+						for _, msg := range msgs {
+							ids = append(ids, msg.ID)
+							last = msg.ID
+						}
+						_, err = cl.XDel(k, ids...).Result()
+						if err != nil {
+							logger.WithError(err).Errorf("Failed to XDEL from stream '%s', continue to next stream", k)
```
Comment (johanstokking) [inline cmd/ttn-lw-stack/commands/redis.go]:
```suggestion
							logger.WithError(err).Errorf("Failed to XDEL from stream %q, continue to next stream", k)
```

**Review by htdvisser — APPROVED** [review]

**Comment by rvolosatovs** [comment]
@johanstokking

**Review by johanstokking — APPROVED** [review]

**Pull request merged**